src/plugins/tonegen/dtmf.c

```diff
@@ -0,0 +1,279 @@
+/*************************************************************************
+This file is part of tone-generator
+
+Copyright (C) 2010 Nokia Corporation.
+
+This library is free software; you can redistribute
+it and/or modify it under the terms of the GNU Lesser General Public
+License as published by the Free Software Foundation
+version 2.1 of the License.
+
+This library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public
+License along with this library; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301
+USA.
+*************************************************************************/
+
+#define _GNU_SOURCE
+
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include <glib.h>
+
+#include <ngf/log.h>
+#include <trace/trace.h>
+
+#include "ausrv.h"
+#include "stream.h"
+#include "tone.h"
+#include "indicator.h"
+#include "dtmf.h"
+#include "dbusif.h"
+
+#define TRACE(f, args...) trace_write(trctx, trflags, trkeys, f, ##args)
+#define LOG_CAT "tonegen-dtmf: "
+
+#define MUTE_ON  1
+#define MUTE_OFF 0
+
+struct dtmf {
+    char           symbol;
+    uint32_t       low_freq;
+    uint32_t       high_freq;
+};
+
+
+static struct dtmf dtmf_defs[DTMF_MAX] = {
+    {'0', 941, 1336},
+    {'1', 697, 1209},
+    {'2', 697, 1336},
+    {'3', 697, 1477},
+    {'4', 770, 1209},
+    {'5', 770, 1336},
+    {'6', 770, 1477},
+    {'7', 852, 1209},
+    {'8', 852, 1336},
+    {'9', 852, 1477},
+    {'*', 941, 1209},
+    {'#', 941, 1477},
+    {'A', 697, 1633},
+    {'B', 770, 1633},
+    {'C', 852, 1633},
+    {'D', 941, 1633}
+};
+
+static char   *dtmf_stream = STREAM_DTMF;
+static void   *dtmf_props  = NULL;
+static int     vol_scale   = 100;
+static guint   mute        = MUTE_OFF;
+static guint   tmute_id;
+
+
+static void destroy_callback(void *);
+static void set_mute_timeout(struct ausrv *, guint);
+static gboolean mute_timeout_callback(gpointer);
+static void request_muting(struct ausrv *, dbus_bool_t);
+
+
+
+int dtmf_init(void)
+{
+    return 0;
+}
+
+void dtmf_play(struct ausrv *ausrv, uint type, uint32_t vol, int dur, const char *extra_properties)
+{
+    struct stream *stream = stream_find(ausrv, dtmf_stream);
+    struct dtmf   *dtmf   = dtmf_defs + type;
+    uint32_t       per    = dur;
+    uint32_t       play   = dur > 60000 ? dur - 20000 : dur;
```

Comment:
What is `60000` and `20000`? Can we maybe put those into named constants? Same for `10000` below.